src/qutip_qip/decompose/decompose_general_qubit_gate.py

```diff
@@ -0,0 +1,932 @@
+    # The input is non-expanded output
+    """Outputs decomposition of mult-qubit controlled gate with last qubit as
+    target by using ancilla qubits.
+
+    First object in two-level array output.
+    """
+    # first decomposition
+    ind_list = _create_list_of_indices_from_two_level_array(
+        two_level_arrays_non_expand)
+    ctrl_targ_inf = _shift_last_qubit_ancilla(ind_list, num_qubits)[0]
+    total_qubits_for_circ = _shift_last_qubit_ancilla(ind_list, num_qubits)[1]
+    first_targ_qubit = ctrl_targ_inf[0][0]['targets ='][0]
+    first_ctrl = len(ctrl_targ_inf[0][0]['controls ='])-1
+
+    last_two_level_array = two_level_arrays_non_expand[0][1].full()
+    first_gate_list = _two_qubit_last_target(
+        last_two_level_array, first_targ_qubit, first_ctrl)
+
+    # decompose A1 controlled on second last qubit
+    ind_A1 = 0
+    # get arg_value to create a rz qobj
+    arg_A1 = first_gate_list[ind_A1].arg_value
+    A1_qobj = rz(arg_A1, N=None, target=first_targ_qubit)
+    A1_gate_list = _two_qubit_last_target(
+        A1_qobj.full(), first_targ_qubit, first_ctrl)
+
+    # decompose A2 controlled on second last qubit
+    ind_A2 = 1
+    # get arg_value to create a ry qobj
+    arg_A2 = first_gate_list[ind_A2].arg_value
+    A2_qobj = ry(arg_A2, N=None, target=first_targ_qubit)
+    A2_gate_list = _two_qubit_last_target(
+        A2_qobj.full(), first_targ_qubit, first_ctrl)
+
+    # decompose B1 controlled on second last qubit
+    ind_B1 = 5
+    # get arg_value to create a ry qobj
+    arg_B1 = first_gate_list[ind_B1].arg_value
+    B1_qobj = ry(arg_B1, N=None, target=first_targ_qubit)
+    B1_gate_list = _two_qubit_last_target(
+        B1_qobj.full(), first_targ_qubit, first_ctrl)
+
+    # decompose B2 controlled on second last qubit
+    ind_B2 = 6
+    # get arg_value to create a rz qobj
+    arg_B2 = first_gate_list[ind_B2].arg_value
+    B2_qobj = rz(arg_B2, N=None, target=first_targ_qubit)
+    B2_gate_list = _two_qubit_last_target(
+        B2_qobj.full(), first_targ_qubit, first_ctrl)
+
+    # decompose C controlled on second last qubit
+    ind_C = 10
+    # get arg_value to create a rz qobj
+    arg_C = first_gate_list[ind_C].arg_value
+    C_qobj = rz(arg_C, N=None, target=first_targ_qubit)
+    C_gate_list = _two_qubit_last_target(
+        C_qobj.full(), first_targ_qubit, first_ctrl)
+    # remove gates at ind_A1,....
+
+    PauliX = first_gate_list[2]
+    CNOT1 = first_gate_list[3]
+
+    toffoli_CNOT_decomposition = _full_CNOT_to_Toffoli(num_qubits-1)
+
+    # full_gate_list = []
+    sub_lists = [
+            A1_gate_list,
+            A2_gate_list,
+            toffoli_CNOT_decomposition, CNOT1, toffoli_CNOT_decomposition,
+            B1_gate_list,  B2_gate_list, toffoli_CNOT_decomposition, CNOT1,
+            toffoli_CNOT_decomposition, C_gate_list, first_gate_list[-1]]
+
+    full_gate_list = list(flatten(sub_lists))
+
+    return(full_gate_list, total_qubits_for_circ)
```

Comment:
output of this function is not correct since ancilla qubits are not added correctly.